File: PregnancyDiary.java
// PregnancyDiary.java
import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.util.*;
import com.google.gson.*;

class Note {
    String id;
    String content;
    String category;
    String date;
    Double weight;
    Integer mood;
    String created_at;

    Note() {}
    Note(String content, String category, String date, Double weight, Integer mood) {
        this.id = UUID.randomUUID().toString().substring(0,8);
        this.content = content;
        this.category = category;
        this.date = date != null ? date : LocalDate.now().toString();
        this.weight = weight;
        this.mood = mood;
        this.created_at = Instant.now().toString();
    }
}

public class PregnancyDiary {
    private List<Note> notes = new ArrayList<>();
    private final String dataFile = "pregnancy_diary.json";
    private final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    public PregnancyDiary() { load(); }

    private void load() {
        try {
            Path path = Paths.get(dataFile);
            if (Files.exists(path)) {
                String json = new String(Files.readAllBytes(path));
                Note[] arr = gson.fromJson(json, Note[].class);
                notes = Arrays.asList(arr);
            }
        } catch (Exception e) {}
    }

    private void save() {
        try {
            Files.write(Paths.get(dataFile), gson.toJson(notes).getBytes());
        } catch (Exception e) {}
    }

    public void add(String content, String category, String date, Double weight, Integer mood) {
        Note n = new Note(content, category, date, weight, mood);
        notes.add(n);
        save();
        System.out.printf("✅ Note added (ID: %s)%n", n.id);
    }

    public void list(String category, String since) {
        List<Note> filtered = notes;
        if (category != null && !category.isEmpty()) {
            filtered = new ArrayList<>();
            for (Note n : notes) if (n.category.equals(category)) filtered.add(n);
        }
        if (since != null && !since.isEmpty()) {
            List<Note> temp = new ArrayList<>();
            for (Note n : filtered) if (n.date.compareTo(since) >= 0) temp.add(n);
            filtered = temp;
        }
        if (filtered.isEmpty()) {
            System.out.println("No matching notes.");
            return;
        }
        System.out.printf("\n📋 Pregnancy Diary (%d entries):\n\n", filtered.size());
        for (Note n : filtered) {
            String weightStr = n.weight != null ? String.format(" – %.1f kg", n.weight) : "";
            String moodStr = n.mood != null ? " ⭐" + "⭐".repeat(n.mood-1) : "";
            System.out.printf("%s: [%s] %s%s%s%n", n.date, n.category, n.content, weightStr, moodStr);
        }
    }

    public void stats() {
        if (notes.isEmpty()) {
            System.out.println("No entries yet.");
            return;
        }
        Map<String, Integer> cats = new HashMap<>();
        List<Integer> moods = new ArrayList<>();
        List<Double> weights = new ArrayList<>();
        for (Note n : notes) {
            cats.put(n.category, cats.getOrDefault(n.category, 0) + 1);
            if (n.mood != null) moods.add(n.mood);
            if (n.weight != null) weights.add(n.weight);
        }
        System.out.println("\n📊 Statistics:");
        System.out.printf("  Total entries: %d%n", notes.size());
        for (Map.Entry<String, Integer> e : cats.entrySet()) {
            System.out.printf("  %s entries: %d%n", e.getKey(), e.getValue());
        }
        if (!moods.isEmpty()) {
            double avg = moods.stream().mapToInt(Integer::intValue).average().orElse(0);
            System.out.printf("  Average mood: %.1f/5%n", avg);
        }
        if (!weights.isEmpty()) {
            System.out.printf("  Latest weight: %.1f kg%n", weights.get(weights.size()-1));
        }
    }

    public void search(String term) {
        List<Note> results = new ArrayList<>();
        for (Note n : notes) {
            if (n.content.toLowerCase().contains(term.toLowerCase())) results.add(n);
        }
        if (results.isEmpty()) {
            System.out.println("No matches.");
            return;
        }
        System.out.printf("\n🔍 Found %d note(s):%n", results.size());
        for (Note n : results) {
            System.out.printf("%s: [%s] %s%n", n.date, n.category, n.content);
        }
    }

    public void exportHTML(String filename) throws IOException {
        StringBuilder html = new StringBuilder();
        html.append("<!DOCTYPE html>\n<html><head><title>Pregnancy Diary</title>\n");
        html.append("<style>body{font-family:sans-serif;margin:30px;background:#f5f5f5;}\n");
        html.append(".note{background:#fff;border-radius:8px;padding:15px;margin:10px 0;box-shadow:0 2px 4px rgba(0,0,0,0.1);}\n");
        html.append(".meta{color:#999;font-size:0.9em;}\n");
        html.append("</style></head><body>\n<h1>🤰 Pregnancy Diary</h1>\n");
        for (int i = notes.size()-1; i >= 0; i--) {
            Note n = notes.get(i);
            html.append("<div class=\"note\">\n");
            html.append("<div class=\"meta\">").append(n.date).append(" – ").append(n.category).append("</div>\n");
            html.append("<div>").append(n.content).append("</div>\n");
            if (n.weight != null) html.append("<div>Weight: ").append(n.weight).append(" kg</div>\n");
            if (n.mood != null) html.append("<div>Mood: ").append("⭐".repeat(n.mood)).append("</div>\n");
            html.append("</div>\n");
        }
        html.append("</body></html>");
        Files.write(Paths.get(filename), html.toString().getBytes());
        System.out.printf("✅ Exported to %s%n", filename);
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.out.println("Usage: PregnancyDiary <command> [options]");
            return;
        }
        PregnancyDiary d = new PregnancyDiary();
        String cmd = args[0];
        Map<String, String> params = new HashMap<>();
        for (int i=1; i<args.length; i++) {
            if (args[i].startsWith("--") && i+1 < args.length) {
                params.put(args[i].substring(2), args[++i]);
            }
        }
        switch (cmd) {
            case "add":
                if (args.length < 2) { System.out.println("add <content> [--category CAT] [--date DATE] [--weight KG] [--mood RATING]"); return; }
                String content = args[1];
                String category = params.getOrDefault("category", "general");
                String date = params.get("date");
                Double weight = params.containsKey("weight") ? Double.parseDouble(params.get("weight")) : null;
                Integer mood = params.containsKey("mood") ? Integer.parseInt(params.get("mood")) : null;
                d.add(content, category, date, weight, mood);
                break;
            case "list":
                d.list(params.get("category"), params.get("since"));
                break;
            case "stats":
                d.stats();
                break;
            case "search":
                if (args.length < 2) { System.out.println("search <term>"); return; }
                d.search(args[1]);
                break;
            case "export":
                String filename = params.getOrDefault("filename", "diary.html");
                d.exportHTML(filename);
                break;
            default:
                System.out.println("Unknown command.");
        }
    }
}

File: pregnancy_diary.go
// pregnancy_diary.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"time"
	"github.com/google/uuid"
)

type Note struct {
	ID        string  `json:"id"`
	Content   string  `json:"content"`
	Category  string  `json:"category"`
	Date      string  `json:"date"`
	Weight    float64 `json:"weight,omitempty"`
	Mood      int     `json:"mood,omitempty"`
	CreatedAt string  `json:"created_at"`
}

type Diary struct {
	Notes []Note `json:"notes"`
}

var dataFile = "pregnancy_diary.json"

func (d *Diary) load() {
	data, err := os.ReadFile(dataFile)
	if err != nil {
		return
	}
	json.Unmarshal(data, d)
}

func (d *Diary) save() {
	data, _ := json.MarshalIndent(d, "", "  ")
	os.WriteFile(dataFile, data, 0644)
}

func (d *Diary) add(content, category, date string, weight float64, mood int) {
	if date == "" {
		date = time.Now().Format("2006-01-02")
	}
	note := Note{
		ID:        uuid.New().String()[:8],
		Content:   content,
		Category:  category,
		Date:      date,
		Weight:    weight,
		Mood:      mood,
		CreatedAt: time.Now().Format(time.RFC3339),
	}
	d.Notes = append(d.Notes, note)
	d.save()
	fmt.Printf("✅ Note added (ID: %s)\n", note.ID)
}

func (d *Diary) list(category, since string) {
	notes := d.Notes
	if category != "" {
		var filtered []Note
		for _, n := range notes {
			if n.Category == category {
				filtered = append(filtered, n)
			}
		}
		notes = filtered
	}
	if since != "" {
		var filtered []Note
		for _, n := range notes {
			if n.Date >= since {
				filtered = append(filtered, n)
			}
		}
		notes = filtered
	}
	if len(notes) == 0 {
		fmt.Println("No matching notes.")
		return
	}
	fmt.Printf("\n📋 Pregnancy Diary (%d entries):\n\n", len(notes))
	for _, n := range notes {
		weightStr := ""
		if n.Weight > 0 {
			weightStr = fmt.Sprintf(" – %.1f kg", n.Weight)
		}
		moodStr := ""
		if n.Mood > 0 {
			moodStr = " ⭐"
			for i := 1; i < n.Mood; i++ {
				moodStr += "⭐"
			}
		}
		fmt.Printf("%s: [%s] %s%s%s\n", n.Date, n.Category, n.Content, weightStr, moodStr)
	}
}

func (d *Diary) stats() {
	if len(d.Notes) == 0 {
		fmt.Println("No entries yet.")
		return
	}
	cats := make(map[string]int)
	var moods []int
	var weights []float64
	for _, n := range d.Notes {
		cats[n.Category]++
		if n.Mood > 0 {
			moods = append(moods, n.Mood)
		}
		if n.Weight > 0 {
			weights = append(weights, n.Weight)
		}
	}
	fmt.Println("\n📊 Statistics:")
	fmt.Printf("  Total entries: %d\n", len(d.Notes))
	for cat, count := range cats {
		fmt.Printf("  %s entries: %d\n", cat, count)
	}
	if len(moods) > 0 {
		sum := 0
		for _, m := range moods {
			sum += m
		}
		avg := float64(sum) / float64(len(moods))
		fmt.Printf("  Average mood: %.1f/5\n", avg)
	}
	if len(weights) > 0 {
		fmt.Printf("  Latest weight: %.1f kg\n", weights[len(weights)-1])
	}
}

func (d *Diary) search(term string) {
	var results []Note
	for _, n := range d.Notes {
		if contains(n.Content, term) {
			results = append(results, n)
		}
	}
	if len(results) == 0 {
		fmt.Println("No matches.")
		return
	}
	fmt.Printf("\n🔍 Found %d note(s):\n", len(results))
	for _, n := range results {
		fmt.Printf("%s: [%s] %s\n", n.Date, n.Category, n.Content)
	}
}

func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(substr) == 0 ||
		(s[0:len(substr)] == substr) || (len(s) > len(substr) && (s[len(s)-len(substr):] == substr ||
		indexOf(s, substr) != -1)))
}

func indexOf(s, substr string) int {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return i
		}
	}
	return -1
}

func (d *Diary) exportHTML(filename string) {
	html := `<!DOCTYPE html>
<html><head><title>Pregnancy Diary</title>
<style>body{font-family:sans-serif;margin:30px;background:#f5f5f5;}
.note{background:#fff;border-radius:8px;padding:15px;margin:10px 0;box-shadow:0 2px 4px rgba(0,0,0,0.1);}
.meta{color:#999;font-size:0.9em;}
</style></head><body>
<h1>🤰 Pregnancy Diary</h1>
`
	for i := len(d.Notes) - 1; i >= 0; i-- {
		n := d.Notes[i]
		html += fmt.Sprintf(`<div class="note">
<div class="meta">%s – %s</div>
<div>%s</div>`, n.Date, n.Category, n.Content)
		if n.Weight > 0 {
			html += fmt.Sprintf("<div>Weight: %.1f kg</div>", n.Weight)
		}
		if n.Mood > 0 {
			html += fmt.Sprintf("<div>Mood: %s</div>", stars(n.Mood))
		}
		html += "</div>"
	}
	html += "</body></html>"
	os.WriteFile(filename, []byte(html), 0644)
	fmt.Printf("✅ Exported to %s\n", filename)
}

func stars(n int) string {
	s := ""
	for i := 0; i < n; i++ {
		s += "⭐"
	}
	return s
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: pregnancy_diary <command> [options]")
		return
	}
	d := &Diary{}
	d.load()
	cmd := os.Args[1]

	switch cmd {
	case "add":
		addCmd := flag.NewFlagSet("add", flag.ExitOnError)
		content := addCmd.String("content", "", "")
		category := addCmd.String("category", "general", "")
		date := addCmd.String("date", "", "")
		weight := addCmd.Float64("weight", 0, "")
		mood := addCmd.Int("mood", 0, "")
		addCmd.Parse(os.Args[2:])
		if *content == "" && len(addCmd.Args()) > 0 {
			*content = addCmd.Args()[0]
		}
		if *content == "" {
			fmt.Println("add requires content")
			return
		}
		d.add(*content, *category, *date, *weight, *mood)

	case "list":
		listCmd := flag.NewFlagSet("list", flag.ExitOnError)
		category := listCmd.String("category", "", "")
		since := listCmd.String("since", "", "")
		listCmd.Parse(os.Args[2:])
		d.list(*category, *since)

	case "stats":
		d.stats()

	case "search":
		if len(os.Args) < 3 {
			fmt.Println("search <term>")
			return
		}
		d.search(os.Args[2])

	case "export":
		exportCmd := flag.NewFlagSet("export", flag.ExitOnError)
		filename := exportCmd.String("filename", "diary.html", "")
		exportCmd.Parse(os.Args[2:])
		d.exportHTML(*filename)

	default:
		fmt.Println("Unknown command.")
	}
}
